python/ray/data/dataset.py

```diff
@@ -2868,75 +2832,72 @@ def to_tf(
             local_shuffle_seed: The seed to use for the local random shuffle.
 
         Returns:
-            A tf.data.Dataset.
+            A ``tf.data.Dataset`` that yields inputs and targets.
         """
-
-        # argument exception checking is done in from_generator
+        from ray.air._internal.tensorflow_utils import get_type_spec
+        from ray.train.tensorflow import prepare_dataset_shard
 
         try:
             import tensorflow as tf
         except ImportError:
             raise ValueError("tensorflow must be installed!")
 
-        from ray.air._internal.tensorflow_utils import convert_pandas_to_tf_tensor
+        if self._dataset_format() == "simple":
+            raise NotImplementedError(
+                "`to_tf` doesn't support simple datasets. Call `map_batches` and "
+                "convert your data to a tabular format. Alternatively, call the more-"
+                "flexible `iter_batches` in place of `to_tf`."
+            )
 
-        # `output_signature` can be a tuple but not a list. See
-        # https://stackoverflow.com/questions/59092423/what-is-a-nested-structure-in-tensorflow.
-        if isinstance(output_signature, list):
-            output_signature = tuple(output_signature)
+        def validate_column(column: str) -> None:
+            valid_columns = self.schema().names
```

Comment:
Could we move this outside of the function scope, so it's only called once instead of `O(columns)` times? It shouldn't be _that_ expensive with `fetch_if_missing=False` (default), but those semantics might change in the near future. https://github.com/ray-project/ray/blob/ed0af4b951f7fb4e3b3421ca4876a45231ca60c4/python/ray/data/_internal/plan.py#L213-L263